File: app/src/main/java/com/kazucocoa/droidtesthelper/HandleAnimations.java
package com.kazucocoa.droidtesthelper;

import android.content.Intent;
import android.os.IBinder;
import android.util.Log;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;

public class HandleAnimations {

    private static String TAG = HandleAnimations.class.getSimpleName();

    private final float DISABLE = 0.0f;

    private final float ENABLE = 1.0f;

    private static String animationExtra = "ANIMATION";

    private Method setAnimationScalesMethod;

    private Method getAnimationScalesMethod;

    private Object windowManagerObject;

    public static boolean hasExtraRegardingAnimation(Intent intent) {
        return intent.hasExtra(animationExtra);
    }

    public static void enableAnimationsWithIntent(Intent intent) {
        boolean enableAnimation = intent.getBooleanExtra(animationExtra, false);
        HandleAnimations handleAnimations = new HandleAnimations();
        handleAnimations.AnimationHandler();

        if (enableAnimation) {
            handleAnimations.enableAnimations();
        } else {
            handleAnimations.disableAnimations();
        }
    }

    public void AnimationHandler() {
        try {
            Class<?> windowManagerStubClazz = Class.forName("android.view.IWindowManager$Stub");
            Method asInterface = windowManagerStubClazz.getDeclaredMethod("asInterface", IBinder.class);

            Class<?> serviceManagerClazz = Class.forName("android.os.ServiceManager");
            Method getService = serviceManagerClazz.getDeclaredMethod("getService", String.class);

            Class<?> windowManagerClazz = Class.forName("android.view.IWindowManager");

            setAnimationScalesMethod = windowManagerClazz.getDeclaredMethod("setAnimationScales", float[].class);
            getAnimationScalesMethod = windowManagerClazz.getDeclaredMethod("getAnimationScales");

            IBinder windowManagerBinder = (IBinder) getService.invoke(null, "window");
            windowManagerObject = asInterface.invoke(null, windowManagerBinder);
        } catch (Exception e) {
            throw new RuntimeException("Failed to access animation methods", e);
        }
    }

    public void disableAnimations() {
        try {
            setAnimationScaleWith(DISABLE);
        } catch (Exception e) {
            Log.e(TAG, "Failed to disable animations", e);
        }
    }

    public void enableAnimations() {
        try {
            setAnimationScaleWith(ENABLE);
        } catch (Exception e) {
            Log.e(TAG, "Failed to enable animations", e);
        }
    }

    private void setAnimationScaleWith(float scaleFactor) throws InvocationTargetException, IllegalAccessException {
        float[] scaleFactors = (float[]) getAnimationScalesMethod.invoke(windowManagerObject);
        Arrays.fill(scaleFactors, scaleFactor);
        setAnimationScalesMethod.invoke(windowManagerObject, new Object[]{scaleFactors});
    }
}

File: droidtesthelperlib/src/main/java/com/kazucocoa/droidtesthelperlib/HandleAnimations.kt
package com.kazucocoa.droidtesthelperlib

import android.content.Intent
import android.os.IBinder
import android.util.Log

import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.Arrays

class HandleAnimations {

    private val DISABLE = 0.0f

    private val ENABLE = 1.0f

    private lateinit var setAnimationScalesMethod: Method

    private lateinit var getAnimationScalesMethod: Method

    private lateinit var windowManagerObject: Any

    private fun animationHandler() {
        try {
            val asInterface = Class.forName("android.view.IWindowManager\$Stub").getDeclaredMethod("asInterface", IBinder::class.java)

            val getService = Class.forName("android.os.ServiceManager").getDeclaredMethod("getService", String::class.java)

            val windowManagerClazz = Class.forName("android.view.IWindowManager")

            setAnimationScalesMethod = Class.forName("android.view.IWindowManager").getDeclaredMethod("setAnimationScales", FloatArray::class.java)
            getAnimationScalesMethod = windowManagerClazz.getDeclaredMethod("getAnimationScales")
            windowManagerObject = asInterface.invoke(null, getService.invoke(null, "window") as IBinder)
        } catch (e: Exception) {
            throw RuntimeException("Failed to access animation methods", e)
        }
    }

    private fun disableAnimations() {
        try {
            setAnimationScaleWith(DISABLE)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to disable animations", e)
        }

    }

    fun enableAnimations() {
        try {
            setAnimationScaleWith(ENABLE)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to enable animations", e)
        }

    }

    @Throws(InvocationTargetException::class, IllegalAccessException::class)
    private fun setAnimationScaleWith(scaleFactor: Float) {
        (getAnimationScalesMethod.invoke(windowManagerObject) as FloatArray).let {
            Arrays.fill(it, scaleFactor)
            setAnimationScalesMethod.invoke(windowManagerObject, it)
        }
    }

    companion object {

        private val TAG = HandleAnimations::class.java.simpleName

        private const val animationExtra = "ANIMATION"

        fun hasExtraRegardingAnimation(intent: Intent): Boolean {
            return intent.hasExtra(animationExtra)
        }

        fun enableAnimationsWithIntent(intent: Intent) {
            val enableAnimation = intent.getBooleanExtra(animationExtra, false)
            val handleAnimations = HandleAnimations()
            handleAnimations.animationHandler()

            if (enableAnimation) {
                handleAnimations.enableAnimations()
            } else {
                handleAnimations.disableAnimations()
            }
        }
    }
}
